File: plugin/test/contract.test.ts
import { describe, expect, it } from "vitest";
import * as fs from "node:fs";
import * as path from "node:path";
import { parseEnvelope } from "../src/utils/validate";

/** Shared contract fixtures — same file the Rust contract test consumes. */
function loadFixtures(): Array<{ name: string; valid: boolean; wire: unknown }> {
  const p = path.resolve(__dirname, "..", "..", "schemas", "protocol", "fixtures.jsonl");
  const raw = fs.readFileSync(p, "utf8");
  return raw
    .split("\n")
    .filter((l) => l.trim().length > 0 && !l.trim().startsWith("#"))
    .map((l) => JSON.parse(l));
}

describe("protocol fixtures (shared with core)", () => {
  it("exist and are meaningful", () => {
    expect(loadFixtures().length).toBeGreaterThanOrEqual(5);
  });

  it("parse exactly when marked valid, fail when marked invalid", () => {
    for (const f of loadFixtures()) {
      const parsed = parseEnvelope(JSON.stringify(f.wire));
      if (f.valid) {
        expect(parsed, `fixture ${f.name} should parse`).not.toBeNull();
      } else {
        expect(parsed, `fixture ${f.name} should be rejected`).toBeNull();
      }
    }
  });
});

File: core/tests/contract_test.rs
//! Contract test: every fixture in schemas/protocol/fixtures.jsonl must parse
//! according to its `valid` flag. Keeps Rust and TypeScript in lockstep.

use serde_json::Value;
use sovereign_core::protocol::Envelope;

fn fixtures_path() -> std::path::PathBuf {
    std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../schemas/protocol/fixtures.jsonl")
}

fn load_fixtures() -> Vec<Value> {
    let raw = std::fs::read_to_string(fixtures_path()).expect("fixtures file exists");
    raw.lines()
        .filter(|l| !l.trim().is_empty() && !l.trim().starts_with('#'))
        .map(|l| serde_json::from_str(l).expect("fixture line is valid JSON"))
        .collect()
}

#[test]
fn all_fixtures_match_rust_protocol_types() {
    let fixtures = load_fixtures();
    assert!(
        fixtures.len() >= 5,
        "expected a meaningful fixture set, got {}",
        fixtures.len()
    );

    for fixture in &fixtures {
        let name = fixture["name"].as_str().unwrap_or("<unnamed>");
        let wire = fixture["wire"].to_string();
        let parsed: Result<Envelope, _> = serde_json::from_str(&wire);
        if fixture["valid"].as_bool().unwrap_or(true) {
            assert!(parsed.is_ok(), "fixture `{name}` should parse: {wire}");
            // Re-serialization must not invent fields.
            let reparsed = serde_json::from_str::<Envelope>(
                &serde_json::to_string(&parsed.unwrap()).unwrap(),
            );
            assert!(reparsed.is_ok(), "fixture `{name}` should round-trip");
        } else {
            assert!(parsed.is_err(), "fixture `{name}` should be rejected");
        }
    }
}
